include/veriblock/blockchain/pop/fork_resolution.hpp

```diff
@@ -39,6 +43,129 @@ struct ComparePopScore {
   int keystoneInterval;
 };
 
+template <typename EndorsementBlockType, typename EndorsementChainParams>
+std::vector<KeystoneContext> getKeystoneContext(
+    const std::vector<ProtoKeystoneContext<EndorsementBlockType>>& chain,
+    const BlockTree<EndorsementBlockType, EndorsementChainParams>&
+        endorsementBlockTree) {
+  std::vector<KeystoneContext> ret;
+
+  std::transform(
+      chain.begin(),
+      chain.end(),
+      std::back_inserter(ret),
+      [&endorsementBlockTree](const ProtoKeystoneContext<BtcBlock>& pkc) {
+        int earliestEndorsementIndex = std::numeric_limits<int32_t>::max();
+        for (const auto* btcIndex : pkc.referencedByBlocks) {
+          if (btcIndex == nullptr) {
+            continue;
+          }
+          auto endorsementIndex = btcIndex->height;
+          if (endorsementIndex >= earliestEndorsementIndex) {
+            continue;
+          }
+
+          if (pkc.timestampOfEndorsedBlock < btcIndex->getBlockTime()) {
+            earliestEndorsementIndex = endorsementIndex;
+            continue;
+          }
+
+          // look at the future BTC blocks and set the--0
+          // earliestEndorsementIndex to a future Bitcoin block
+          auto btcBest = endorsementBlockTree.getBestChain();
+          for (int adjustedEndorsementIndex = endorsementIndex + 1;
+               adjustedEndorsementIndex <= btcBest.chainHeight();
+               adjustedEndorsementIndex++) {
+            // Ensure that the keystone's block time isn't later than the
+            // block time of the Bitcoin block it's endorsed in
+            auto* index = btcBest[adjustedEndorsementIndex];
+            if (index == nullptr) {
+              throw std::logic_error("unexpected nullptr in btc best chain");
+            }
+            if (pkc.timestampOfEndorsedBlock < index->getBlockTime()) {
+              // Timestamp of VeriBlock block is lower than Bitcoin block,
+              // set this as the adjusted index if another lower index has
+              // not already been set
+              if (adjustedEndorsementIndex < earliestEndorsementIndex) {
+                earliestEndorsementIndex = adjustedEndorsementIndex;
+              }
+
+              // Always break; we found a valid Bitcoin index and any
+              // future adjustedEndorsementIndex is going to be higher
+              break;
+            }  // end if
+          }    // end for
+        }      // end for
```

Comment:
Could we split this code to smaller function? At least move it out of the lambda function.